Unbound functionality in ICMsg

doki-nordic opened the pull request:
In some cases, CPUs that may need to reset or temporary stop communication. This commit adds "unbound" functionality that provides a callback to IPC service user when connection was interrupted for some reason, e.g. expected or unexpected CPU reset, closing the endpoint. The "unbound" callback is optional to implement by endpoints. This commit implements it in the ICMsg backend.

The PR also adds `ipc_sessions` test that checks if `ipc_service` is able to correctly handle this feature. The only backend that is tested is ICMsg since only it implements this feature.

rakons commented on subsys/ipc/ipc_service/lib/pbuf.c:
```diff
@@ -176,6 +179,44 @@ int pbuf_write(struct pbuf *pb, const char *data, uint16_t len)
+	*buf = &pb->cfg->data_loc[PBUF_PACKET_LEN_SZ];
+	*len = plen;
+
+	sys_cache_data_invd_range((void *)buf, plen);
```
```suggestion
	sys_cache_data_invd_range((void *)*buf, plen);
```

doki-nordic committed:
ipc_service: icmsg: Add "unbound" functionality

In some cases, CPUs that may need to reset or
temporary stop communication. This commit adds "unbound"
functionality that provides a callback to IPC service user
when connection was interrupted for some reason, e.g.
expected or unexpected CPU reset, closing the
endpoint. The "unbound" callback is optional to implement
by endpoints. This commit implements it in the ICMsg
backend.

Signed-off-by: Dominik Kilian <Dominik.Kilian@nordicsemi.no>

doki-nordic committed:
tests: ipc_service: Test restarting session in different scenarios

This commit adds a test that checks if disconnecting
and restarting the IPC session works correctly. The
test is also focused on the "unbound" callback.

Signed-off-by: Dominik Kilian <Dominik.Kilian@nordicsemi.no>
Co-authored-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

rakons commented on include/zephyr/ipc/pbuf.h:
```diff
@@ -140,9 +148,11 @@ struct pbuf {
- * @param dcache_align	Data cache line size.
+ * @param dcache_align		Data cache line size.
+ * @param use_handshake		Add handshake word inside shared memory that can be access with
+ *				@ref pbuf_handshake_read and @ref pbuf_handshake_write.
```
Missing documentation for "compatibility" argument.

rakons commented on include/zephyr/ipc/pbuf.h:
```diff
@@ -223,6 +235,40 @@ int pbuf_write(struct pbuf *pb, const char *buf, uint16_t len);
+ * The pb must be defined with "PBUF_DEFINE" with "use_handshake" set.
+ *
+ * @param pb		A buffer from which data will be read.
+ * @retval uint32_t	The handshake word value.
```
uint32_t ?

rakons reviewed (approved):
Approved with some small style comments. Not a blocker.
